prefetch.bs

```diff
@@ -432,6 +436,8 @@ Update all creation sites to supply an empty string, except for any in this docu
                 1. Let |coopEnforcementResult| be the result of [=creating a cross-origin opener policy enforcement result for navigation=] given <var ignore>navigable</var>'s [=navigable/active document=] and <var ignore>entry</var>'s [=session history entry/document state=]'s [=document state/initiator origin=].
                 1. Set <var ignore>navigationParams</var> to the result of [=creating navigation params by fetching=] given |request|, <var ignore>entry</var>, |coopEnforcementResult|, <var ignore>navigable</var>, <var ignore>sourceSnapshotParams</var>, <var ignore>targetSnapshotParams</var>, <var ignore>cspNavigationType</var>, <var ignore>navigationId</var>, and <var ignore>navTimingType</var>.
 
+    The |sourceDocument| argument should be plumbed from <a spec=HTML>navigate</a>'s |sourceDocument| and <a spec=HTML>traverse the history by a delta</a>'s |sourceDocument|.
+    <p class="issue">How about other callers, i.e. <a spec=HTML>apply pending history changes</a>, <a spec=HTML>finalize a same-document navigation</a> and <a spec=HTML>finalize a cross-document navigation</a>?</p>
```

Comment:
- apply pending history changes: only if we want to support prefetching reloads. Which I'm not sure we do?
- finalize a same-document navigation: no, prefetching is not relevant there
- finalize a cross-document navigation: yes, plumbing it through here is necessary for the path navigate -> finalize a cross-document navigation -> apply the history step.

Reply:
> support prefetching reloads

There is a WPT already for reload (https://github.com/WICG/nav-speculation/pull/212#discussion_r1035155217), but maybe just to check the Chromium behavior?
https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/web_tests/external/wpt/speculation-rules/prefetch/prefetch-traverse-reload.sub.html

> navigate -> finalize a cross-document navigation -> apply the history step

Do you have examples/test cases where prefetched results would be used in this code path? (I'm not familiar with how "finalize a cross-document navigation" works)

Reply:
> There is a WPT already for reload ([#212 (comment)](https://github.com/WICG/nav-speculation/pull/212#discussion_r1035155217)), but maybe just to check the Chromium behavior?

Ah OK, I am happy to keep it then.

> Do you have examples/test cases where prefetched results would be used in this code path? (I'm not familiar with how "finalize a cross-document navigation" works)

This is the most basic code path. E.g. this is the code path which clicking a link or doing `location.href` goes through.

Reply:
https://html.spec.whatwg.org/multipage/browsing-the-web.html#reload says:

> It is intentional that the resulting call to [apply the history step](https://html.spec.whatwg.org/multipage/browsing-the-web.html#apply-the-history-step) does not pass [sourceSnapshotParams](https://html.spec.whatwg.org/multipage/browsing-the-web.html#apply-history-step-source-snapshot) or [initiatorToCheck](https://html.spec.whatwg.org/multipage/browsing-the-web.html#apply-history-step-initiator).

So probably we shouldn't pass `sourceDocumentForPrefetches` as well?

Also, how about passing `sourceDocumentForPrefetches` (or its prefetch records) via `sourceSnapshotParams`?

Pros: It might be consistent with the existing decisions whether we should pass sourceDocument-related informatio n (`sourceSnapshotParams` or `sourceDocumentForPrefetches`) or not

Cons?: this means not passing `sourceDocumentForPrefetches` in `finalize a cross-document navigation`. IIUC `sourceDocumentForPrefetches` will be still passed for ordinal cross-document navigations via navigate -> [attempt to populate the history entry's document](https://html.spec.whatwg.org/multipage/browsing-the-web.html#attempt-to-populate-the-history-entry's-document) -> [creating navigation params by fetching](https://html.spec.whatwg.org/multipage/browsing-the-web.html#create-navigation-params-by-fetching), even if we don't pass `sourceDocumentForPrefetches` via navigate -> `finalize a cross-document navigation` though (I'm not sure the differences between the two paths, but probably the path via [attempt to populate the history entry's document](https://html.spec.whatwg.org/multipage/browsing-the-web.html#attempt-to-populate-the-history-entry's-document) is the primary one?).

Reply:
Sorry for the delay; I got sick :(.

> So probably we shouldn't pass `sourceDocumentForPrefetches` as well?

Hmm, I'm not sure. Mostly, sourceSnapshotParams and initiatorToCheck are used for security checks which (for some reason) browsers decide to not perform on reloads. The only thing that's kind of similar to how sourceDocumentForPrefetches is used is sourceSnapshotParams's fetch client, which is used to determine referrer and such. (E.g. for reloads, it always looks like the page is reloading itself, according to the `Referer` header.)

> Cons?: this means not passing `sourceDocumentForPrefetches` in `finalize a cross-document navigation`. IIUC `sourceDocumentForPrefetches` will be still passed for ordinal cross-document navigations via navigate -> [attempt to populate the history entry's document](https://html.spec.whatwg.org/multipage/browsing-the-web.html#attempt-to-populate-the-history-entry's-document) -> [creating navigation params by fetching](https://html.spec.whatwg.org/multipage/browsing-the-web.html#create-navigation-params-by-fetching), even if we don't pass `sourceDocumentForPrefetches` via navigate -> `finalize a cross-document navigation` though (I'm not sure the differences between the two paths, but probably the path via [attempt to populate the history entry's document](https://html.spec.whatwg.org/multipage/browsing-the-web.html#attempt-to-populate-the-history-entry's-document) is the primary one?).

Sorry, I realize I was misleading in what I said earlier. The paths are:

- push/replace navigation: `navigate` -> `attempt to populate` -> `finalize a cross-document navigation` -> `apply the history step`. In `apply the history step`, targetEntry's document will be present, because of the earlier `attempt to populate` step.

- reload navigation: `reload` -> `apply pending history changes` -> `apply the history step` -> `attempt to populate`. This time we reach `attempt to populate` from `apply the history step`, because targetEntry's document state's reload pending is true 

- back/forward navigation, bfcache: `traverse the history by a delta` -> `apply the history step`. In `apply the history step`, targetEntry's document will be present, because bfcache.

- back/forward navigation, non-bfcache: `traverse the history by a delta` -> `apply the history step` -> `attempt to populate`. This time we reach `attempt to populate` from `apply the history step`, because targetEntry's document is null, because no bfcache.

Does that make sense? Feel free to grab me in person or in realtime if it doesn't; I know this stuff is complicated and I don't want to be confusing about it again.

So, I think your idea of putting sourceDocumentForPrefetches or the prefetch records in source snapshot params mostly works, and feels elegant. The consequence would be that reloads can only use prefetch records from their own document though, e.g. if you had this setup:

- Parent loads `a.html` in an iframe
- Later, parent prefetches `a.html`
- Parent calls `frames[0].reload()`

then the prefetch would not be used. Maybe that's fine, because it's congruent with how the fetch client is set to `frames[0]` instead of to the parent?

Another elegant part of putting the prefetch records (not sourceDocumentForPrefetches) in source snapshot params is that if we start storing the prefetch records on a wider scope, per https://github.com/WICG/nav-speculation/issues/266, it will not change very much: it will just change the snapshotting call to grab them from a different place.

Reply:
Hmm, thank you for inputs and clarifications! Then I'll draft the spec to put prefetch records to source snapshot params.

I'm also less sure about reload's behavior here, but I feel we can prioritize (at least for now) the elegance/consistency of the spec architecture by putting the prefetch records to source snapshot params, and then later (after #266) we can revisit the reload behavior if actually needed in the wild.

If we make prefetch records scoped by network partition keys, I imagine prefetch records would be partitioned by the result of `determine the network partition key given request’s client` (because `sourceDocumentForPrefetches` basically corresponds to source snapshot params' client), while HTTP caches for navigation requests are partitioned by the result of `determine the network partition key given request’s reserved client`, which is perhaps simple.